packages/lit-ssr/src/lib/render-lit-html.ts

```diff
@@ -0,0 +1,551 @@
+ * to run and be set on the instance before we render the element's final
+ * attributes.
+ */
+type CustomElementAttributesOp = {
+  type: 'custom-element-attributes';
+};
+
+/**
+ * Operation to render a custom element's children, usually its shadow root.
+ */
+type CustomElementChildrenOp = {
+  type: 'custom-element-children';
+};
+
+/**
+ * Operation to close a custom element so that its no longer available for
+ * bindings.
+ */
+type CustomElementClosedOp = {
+  type: 'custom-element-close';
+};
+
+type Op =
+  | TextOp
+  | NodePartOp
+  | AttributePartOp
+  | CustomElementOpenOp
+  | CustomElementAttributesOp
+  | CustomElementChildrenOp
+  | CustomElementClosedOp;
+
+const getTemplate = (result: TemplateResult) => {
+  const template = templateCache.get(result.strings);
+  if (template !== undefined) {
+    return template;
+  }
+  const {html, attrNames} = getTemplateHtml(result.strings, result._$litType$);
+  const ast = parseFragment(html, {
+    sourceCodeLocationInfo: true,
+  }) as DefaultTreeDocumentFragment;
+
+  const ops: Array<Op> = [];
+
+  /* The last offset of html written to the stream */
+  let lastOffset: number | undefined = 0;
+
+  /**
+   * Sets `lastOffset` to `offset`, skipping a range of characters. This is
+   * useful for skipping <slot>s and distributed nodes in flattened mode, or
+   * skipping and re-writting lit-html marker nodes.
+   */
+  const skipTo = (offset: number) => {
+    if (lastOffset === undefined) {
+      throw new Error('lastOffset is undefined');
+    }
+    if (offset < lastOffset) {
+      throw new Error(`offset must be greater than lastOffset.
+        offset: ${offset}
+        lastOffset: ${lastOffset}
+      `);
+    }
+    lastOffset = offset;
+  };
+
+  /**
+   * Creates or appends to a text opcode with a substring of the html from the
+   * `lastOffset` flushed to `offset`.
+   */
+  const flushTo = (offset?: number) => {
+    if (lastOffset === undefined) {
+      throw new Error('lastOffset is undefined');
+    }
+    const previousLastOffset = lastOffset;
+    lastOffset = offset;
+    const value = html.substring(previousLastOffset, offset);
+    const op = getLast(ops);
+    if (op !== undefined && op.type === 'text') {
+      op.value += value;
+    } else {
+      ops.push({
+        type: 'text',
+        value,
+      });
+    }
+  };
+
+  const flush = (value: string) => {
+    const op = getLast(ops);
+    if (op !== undefined && op.type === 'text') {
+      op.value += value;
+    } else {
+      ops.push({
+        type: 'text',
+        value,
+      });
+    }
+  };
+
+  // Depth-first node index. Initialized to -1 so that the first child node is
+  // index 0, to match client-side lit-html.
+  let nodeIndex = -1;
+
+  traverse(ast, {
+    pre(node, parent) {
+      if (isCommentNode(node)) {
+        if (node.data === markerMatch) {
+          flushTo(node.sourceCodeLocation!.startOffset);
```

Comment:
Add comment about the `ast` structure. What is `sourceCodeLocation.start/endOffset`?

Reply:
Hmm, not sure how to add much in the way of comments since the code here seems pretty self explanatory: it's the location in the source html string of the given AST node:

223: The current AST node is a comment node
224: The comment node's content matches marker text
225: Flush our string accumulator up to the _start offset_ of the _source code location_ for this comment _node_

Suggestions?

Reply:
`sourceCodeLocation.start/endOffset` are Parse5 defined properties. I thought this part was pretty self-documenting.

Reply:
ok